File: projects/advent_of_code/solutions/year2015/day01/go/solution.go
package main

import (
	"fmt"

	"github.com/jomcgi/homelab/projects/advent_of_code/pkg/aoc"
)

func main() {
	input, err := aoc.ReadFile("../input.txt")
	if err != nil {
		panic(err)
	}

	fmt.Println("Part 1:", part1(input))
	fmt.Println("Part 2:", part2(input))
}

func part1(input string) int {
	floor := 0
	for _, char := range input {
		if char == '(' {
			floor++
		} else if char == ')' {
			floor--
		}
	}
	return floor
}

func part2(input string) int {
	floor := 0
	for position, char := range input {
		if char == '(' {
			floor++
		} else if char == ')' {
			floor--
		}
		if floor == -1 {
			return position + 1
		}
	}
	return 0
}

File: projects/advent_of_code/solutions/year2015/day01/python/solution.py
#!/usr/bin/env python3
"""Advent of Code 2015 Day 1"""

from advent_of_code.python.aoc import lines, ints, read_file


def part1(input_text: str) -> int:
    floor = 0
    for char in input_text.strip():
        if char == "(":
            floor += 1
        elif char == ")":
            floor -= 1
    return floor


def part2(input_text: str) -> int:
    floor = 0
    for position, char in enumerate(input_text.strip(), start=1):
        if char == "(":
            floor += 1
        elif char == ")":
            floor -= 1
        if floor == -1:
            return position
    return 0


if __name__ == "__main__":
    input_text = read_file("../input.txt")

    print(f"Part 1: {part1(input_text)}")
    print(f"Part 2: {part2(input_text)}")
